hibernate-core/src/main/java/org/hibernate/id/IdentifierGeneratorHelper.java

```diff
@@ -134,32 +138,45 @@ else if ( clazz == BigDecimal.class ) {
 			}
 		}
 		else {
-			if ( clazz == Long.class ) {
-				return rs.getLong( identifier );
-			}
-			else if ( clazz == Integer.class ) {
-				return rs.getInt( identifier );
-			}
-			else if ( clazz == Short.class ) {
-				return rs.getShort( identifier );
-			}
-			else if ( clazz == String.class ) {
-				return rs.getString( identifier );
+			try {
+				return extractIdentifier( rs, identifier, type, clazz );
 			}
-			else if ( clazz == BigInteger.class ) {
-				return rs.getBigDecimal( identifier ).setScale( 0, BigDecimal.ROUND_UNNECESSARY ).toBigInteger();
-			}
-			else if ( clazz == BigDecimal.class ) {
-				return rs.getBigDecimal( identifier ).setScale( 0, BigDecimal.ROUND_UNNECESSARY );
-			}
-			else {
-				throw new IdentifierGenerationException(
-						"unrecognized id type : " + type.getName() + " -> " + clazz.getName()
-				);
+			catch (SQLException e) {
+				if(Identifier.isQuoted( identifier )) {
```

Comment:
I checked it and the `Identifier#isQuoted` verifies all the currently supported quoting characters: `, [, and ". As far as I know, there cannot be a 'name' because there is no Dialect to use such a quoting syntax. More, the single quote syntax is reserved for Strings in SQL.